### ISSUE TITLE
Lifecycle issue with RETAIN_DETACH-enabled ViewPager host controller and child controllers

### ISSUE THREAD

**Issue by dimsuz** [bluelinelabs/Conductor#314]
When host controller of `RouterPagerAdapter` has `RETAIN_DETACH` mode set and another controller is being pushed on top of the host controller, then `onDetach` gets called on individual child page controllers inside the adapter, but after popping back to host controller, `onAttach` never gets called on those page controllers. This breaks `onAttach`/`onDetach` symmetry and I suspect is wrong. At least I have problems with this while using a `mosby-conductor` library (it never reattaches view back to a presenter)

I prepared a [repository clone](https://github.com/dimsuz/Conductor/tree/router_pager_no_attach_bug) with a slightly modified `demo` app which can be used to reproduce the problem.
The diff of my changes can be seen in [this commit](https://github.com/dimsuz/Conductor/commit/8cf2d884e83b9061415a85ac6fb311293f05cc17).

Here are the steps to reproduce with this demo app:

1. Launch the demo
2. Go to `ViewPager` menu item
3. Click on the page title (it'll launch a `TextController` on top of the parent (host) controller)
4. Observe the logs:
```
Child Controller Child #0 (Swipe to see more): onDetach()
Child Controller Child #0 (Swipe to see more): onDestroyView()
```
5. Press "BACK" button to pop `TextController` and return to `PagerController`
6. Observe that logs have no calls to `onAttach` or `onCreateView` for `ChildController`...

If you remove `RETAIN_DETACH` mode then everything starts to work and `onAttach` will appear in logs on step 6 above.
Note that `RETAIN_DETACH` is set only on host controller (`PagerController`), child controllers have default behavior, but still their lifecycle ends up being broken.

**Comment by dimsuz** [comment]
I seem to have found a reason why this happens.

When `PagerController` is set to use the `RETAIN_DETACH` mode, its view doesn't get destroyed, but its **child** controllers do get their views destroyed. Child controllers in this case happen to be controllers of pages which reside in `ControllerPagerAdapter`.

So when pushing a new Controller on top of `PagerController` with `RETAIN_DETACH` we get this picture:

`PagerController::onDestroyView` - not called
`ChildController::onDestroyView` (pages) - **called**

so this is kinda asymmetric. parent lives while its children die :)
And also later children will never get attached back. I think this is because parent view wasn't destroyed, so it doesn't care to reattach children, but i'm not sure.

One fix that comes to mind is to set `RETAIN_DETACH` on all child controllers once it has been set on the parent. In `ViewPager`'s case this will need to be done dynamically though - as pages are created...

P.S. I updated my fork to use the latest upstream `develop` branch if you want to play with this. Reproduction steps are still valid.

**Comment by PaulWoitaschek** [comment]
I just ran into this issue too.

RootRouter
--AController(RETAIN_DETACH)
  --AChildController(RELEASE_DETACH)

Now whan I push a `BController` to RootRouter and pop it, I'm back in `AController`, but `AChildController` is blank because its `onDestroyView` was called but no further `onCreateView`.

**Comment by dimsuz** [comment]
Any thoughts on how to approach fixing this? I'd try but I'm not sure what would be the best way. In my previous comment I suggested one, but it doesn't seem a very good one (considering the need to somehow dynamically adjust RETAIN_DETACH-ness in the ViewPager case).

**Comment by mingdong-recolor** [comment]
Any estimate on when this issue will be resolved?

**Comment by EricKuck** [comment]
Fixed via #571
